packages/contracts/contracts/SmartInvoice.sol

```diff
@@ -130,6 +139,43 @@ contract SmartInvoice is
         emit Register(_client, _provider, amounts);
     }
 
+    function extendTerminationTime(uint256 _time) public {
+        require(!locked, "locked");
+        require(_msgSender() == client, "msg.sender !client");
+        terminationTime += _time;
+
+        emit TerminationExtension(_msgSender(), _time);
+    }
+
+    function addMilestones(uint256[] calldata _milestones, uint256 _time)
+        external
+    {
+        require(!locked, "locked");
+        require(_msgSender() == client, "msg.sender !client");
+        require(_milestones.length > 0, "no milestones are being added");
+        require(_milestones.length <= 10, "only 10 new milestones at a time");
+
+        if (_time > 0) extendTerminationTime(_time);
+
+        uint256 newLength = amounts.length + _milestones.length;
+        uint256[] memory baseArray = new uint256[](newLength);
+
+        for (uint256 i = 0; i < amounts.length; i++) {
+            baseArray[i] = amounts[i];
+        }
+        for (uint256 i = amounts.length; i < newLength; i++) {
+            baseArray[i] = _milestones[i - amounts.length];
+        }
+        amounts = baseArray;
+
+        if (_time > 0) emit TerminationExtension(client, _time);
+        emit MilestonesAdded(client, address(this), _milestones);
+    }
+
+    function getMilestones() public view returns (uint256[] memory) {
+        return amounts;
+    }
```

Comment:
This naming sounds odd as it is getting milestones but returning amounts.

Reply:
will change